tsl/src/chunk.c

```diff
@@ -259,3 +260,92 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type)
 
 	return num_results;
 }
+
+/*
+ * drop_chunk_replica:
+ *
+ * This function drops a chunk on a specified data node. It then
+ * removes the metadata about the association of the chunk to this
+ * datanode on the access node.
+ */
+Datum
+drop_chunk_replica(PG_FUNCTION_ARGS)
+{
+	const char *node_name = PG_ARGISNULL(0) ? NULL : NameStr(*PG_GETARG_NAME(0));
+	Oid chunk_relid = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	List *data_nodes, *replicas;
+	ForeignServer *server;
+	Chunk *chunk;
+	ChunkDataNode *cdn;
+	char *drop_cmd;
+
+	if (!OidIsValid(chunk_relid))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid chunk specification %u", chunk_relid)));
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, false);
+
+	if (NULL == chunk)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("Invalid chunk specified")));
+
+	/* It has to be a foreign table chunk */
+	if (chunk->relkind != RELKIND_FOREIGN_TABLE)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("\"%s\" is not a valid remote chunk", get_rel_name(chunk_relid))));
+
+	server = data_node_get_foreign_server(node_name, ACL_USAGE, true, false);
+	Assert(NULL != server);
+
+	/* Early abort on missing permissions */
+	ts_hypertable_permissions_check(chunk_relid, GetUserId());
+
+	/* check that the chunk is indeed present on the specified data node */
+	cdn = ts_chunk_data_node_scan_by_chunk_id_and_node_name(chunk->fd.id,
+															server->servername,
+															CurrentMemoryContext);
+
+	if (!cdn)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("Chunk \"%s\" does not exist on data node \"%s\"",
+						get_rel_name(chunk_relid),
+						node_name)));
+
+	/* Ensure that this chunk remains replicated so as to avoid data loss */
+	replicas = ts_chunk_data_node_scan_by_chunk_id(chunk->fd.id, CurrentMemoryContext);
+
+	/*
+	 * There should be atleast one surviving replica after the deletion here.
+	 *
+	 * We could fetch the corresponding hypertable and check its
+	 * replication_factor. But the user of this function is using it
+	 * to move chunk from one datanode to another and is well aware of
+	 * the replication_factor requirements
+	 */
+	if (list_length(replicas) <= 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_TS_INSUFFICIENT_NUM_DATA_NODES),
+				 errmsg("dropping last chunk could lead to data loss")));
```

Comment:
I suggest to give an error in the error message and the current text can go into details instead. For example:
```suggestion
				 errmsg("cannot drop the last chunk replica"),
				 errdetail("Dropping the last chunk  replica could lead to data loss.")));
```

Reply:
Fine.. changed